external/java/src/main/java/ch/epfl/dedis/skipchain/SkipchainRPC.java

```diff
@@ -74,39 +84,114 @@ public SkipBlock getSkipblock(SkipblockId id) throws CothorityCommunicationExcep
+     *
+     * @return the latest skipblock
+     * @throws CothorityCommunicationException if something goes wrong with communication
+     * @throws CothorityCryptoException if the verification goes wrong
+     */
+    public SkipBlock getLatestSkipblock() throws CothorityCommunicationException, CothorityCryptoException {
+        return this.getLatestSkipblock(this.scID);
+    }
 
-        ByteString msg = roster.sendMessage("Skipchain/GetUpdateChain",
-                request);
+    /**
+     * Returns the latest block and verify that the links are correct. The verification is performed from trustedLatest.
+     *
+     * @param trustedLatest is the latest block ID that the caller trusts, which serves as the source for verification.
+     * @return the latest skipblock
+     * @throws CothorityCommunicationException if something goes wrong with communication
+     * @throws CothorityCryptoException if the verification goes wrong
+     */
+    public SkipBlock getLatestSkipblock(SkipblockId trustedLatest) throws CothorityCommunicationException, CothorityCryptoException {
+        List<SkipBlock> update = new ArrayList<>();
+        for (;;) {
+            // make the request
+            SkipchainProto.GetUpdateChainReply r2;
+            try {
+                SkipchainProto.GetUpdateChain request =
+                        SkipchainProto.GetUpdateChain.newBuilder()
+                                .setLatestID(ByteString.copyFrom(trustedLatest.getId()))
+                                .build();
+                ByteString msg = roster.sendMessage("Skipchain/GetUpdateChain",
+                        request);
+                r2 = SkipchainProto.GetUpdateChainReply.parseFrom(msg);
+            } catch (InvalidProtocolBufferException e) {
+                throw new CothorityCommunicationException(e);
+            }
 
-        try {
-            SkipchainProto.GetUpdateChainReply reply =
-                    SkipchainProto.GetUpdateChainReply.parseFrom(msg);
-            //TODO: add verification that the skipblock is valid by hashing and comparing to the id
+            SkipBlock start = new SkipBlock(r2.getUpdateList().get(0));
+            if (!trustedLatest.equals(start.getId())) {
+                throw new CothorityCryptoException("first returned block does not match requested hash");
+            }
 
-            if (reply.getUpdateCount() == 0){
-                logger.info("didn't find any updates to {}", scID);
-                return null;
+            // Step through the returned blocks one at a time, verifying
+            // the forward links, and that they link correctly backwards.
+            for (int j = 0; j < r2.getUpdateCount(); j++) {
+                SkipBlock b = new SkipBlock(r2.getUpdateList().get(j));
+                if (j == 0 && update.size() > 0) {
+                    if (Arrays.equals(update.get(update.size()-1).getHash(), b.getHash())) {
```

Comment:
I think this line says, "if we are processing the first block, and we've already processed some blocks, then make sure this first block is the same as the last block we've already accepted".

If I'm correct, this is what we want to be doing, so good on ya mate. So add a comment saying so.

If I'm not correct, it means this code is too confusing and needs to be either fixed or documented.

Reply:
this part is a copy/paste from the go code and it's what I understood too. will add a comment to explain